keps/sig-apps/1847-autoremove-statefulset-pvcs/README.md

```diff
@@ -0,0 +1,508 @@
+       action is taken with respect to the PVCs created by the StatefulSet.
+   * `DeleteOnStatefulSetDeletion` - PVCs corresponding to the StatefulSet are deleted when StatefulSet
+       themselves get deleted. When Pods are deleted as part of a scale down, PVCs are not
+       deleted. Thus there may be PVCs owned by the StatefulSet that are not attached to a Pod.
+   * `DeleteOnScaledown` - When a pod is deleted on scale down, the corresponding PVC is deleted as well. 
+       On a scale down followed by a scale up, will wait until the old PVC for the removed Pod is deleted and ensure 
+       that the PVC used is a freshly created one. This policy also implies the
+       former, that is, PVCs will also be deleted when the StatefulSet is deleted.
+2. Add `patch` to the statefulset controller rbac cluster role for `persistentvolumeclaims`.
+
+### User Stories
+
+#### Story 1
+The user is running a StatefulSet as part of an application with a finite lifetime. During
+the application's existence the StatefulSet maintains per-pod state, even across scale-up
+and scale-down. In order to maximize performance, volumes are retained during scale-down
+so that scale-up can leverage the existing volumes. When the application is finished, the
+volumes created by the StatefulSet are no longer needed and can be automatically
+reclaimed.
+
+The user would set the `PersistentVolumeClaimDeletePolicy` as `DeleteOnStatefulSetDelete`
+which would ensure that the PVCs created automatically during the StatefulSet activation
+is deleted once the StatefulSet is deleted.
+
+#### Story 2
+User is very cost conscious and can sustain slower scale-up speeds, even after a
+scale-down. The user does not want to pay for volumes that are not in use, and
+so wants them to be reclaimed as soon as possible, including during
+scale-down. On scale-up a new volume will be provisioned and the new pod will
+have to re-intitialize. However, for short-lived interruptions when a pod is
+killed & recreated, like a rolling update or node disruptions, the data on
+volumes is persisted. This is a key property that ephemeral storage, like
+emptyDir, cannot provide.
+
+User would set the `PersistentVolumeClaimDeletePolicy` as `DeleteOnScaledown` ensuring
+PVCs are deleted when corresponding Pods are deleted. New Pods created during scale-up
+followed by a scale-down will wait for freshly created PVCs.
+
+### Notes/Constraints/Caveats (optional)
+
+This feature applies to PVCs which are dynamically provisioned from the
+volumeClaimTemplate of a StatefulSet. Any PVC and PV provisioned from this
+mechanism will function with this feature. Note that to current knowledge these
+PVCs are identified by the static naming scheme used by StatefulSets. If a user
+pre-provisions a PVC with the same name, it will be treated as an
+auto-provisioned one and used accordingly. It is not clear what happens today if
+a user has pre-provisioned a PVC matching one that the StatefulSet will try to
```

Comment:
Pre-creating a PVC with the Statefulset pvc template naming syntax is a valid and supported operation. The StatefulSet controller won't create the PVC in that case since one already exists.

Reply:
Ah, thank you.

I've clarified this. I think that the best thing to do is have clear behavior about what happens based on the state of the world, and to not try to infer semanics based on the history of how things got created. That seems to be the most k8s/reconciling way to do it IMHO at least.